core/Access.php

```diff
@@ -155,8 +156,24 @@ public function reloadAccess(Auth $auth = null)
             return false;
         }
 
+        $result = null;
+
+        $forceApiSession = Common::getRequestVar('force_api_session', 0, 'int', $_POST);
+        if ($forceApiSession && Piwik::getModule() === 'API' && (Piwik::getAction() === 'index' || !Piwik::getAction())) {
+            $tokenAuth = Common::getRequestVar('token_auth', '', 'string', $_POST);
+            if (!empty($tokenAuth)) {
+                Session::start();
+                $auth = StaticContainer::get(SessionAuth::class);
+                $auth->setTokenAuth($tokenAuth);
+                $result = $auth->authenticate();
+                // if not successful, we will fallback to regular auth
+            }
+        }
```

Comment:
Is this for using the session during API requests? I remember when changing to sessions we didn't want to do this for some reason, do you remember why? Wondering if it's still relevant. It might've been for performance.

Reply:
@diosmosis this is only for requests initiated by the UI. A standard API request will still not use the session. Was like the only workaround I could think of that doesn't break too much. It's actually not even so much of a tokenAuth here but more of an nonce.

Note to myself: Add BruteForce logic here.